File: src/pinusdb/connector/PDB.DotNetSDK/PDBConnection.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Net.Sockets;

namespace PDB.DotNetSDK
{
  public class PDBConnection : IDisposable
  {
    private TcpClient sock;

    private const int kUserNameLen = 48;
    private const int kPwdLen = 4;

    /// <summary>
    /// PinDB连接对象---构造函数
    /// </summary>
    /// <param name="connStr">连接字符串，如：server=127.0.0.1;port=4517;username=sa;password=pinusdb</param>
    public PDBConnection(string connStr)
    {
      this.Port = 0;
      string[] strArr = connStr.Split(new char[] { ';' }, StringSplitOptions.RemoveEmptyEntries);

      foreach (string str in strArr)
      {
        string[] itemArr = str.Split(new char[] { '=' }, StringSplitOptions.RemoveEmptyEntries);
        if (itemArr.Count() != 2)
        {
          throw new PDBException(PDBErrorCode.PdbE_INVALID_PARAM);
        }

        if (itemArr[0].Trim().ToLower().Equals("server"))
        {
          HostName = itemArr[1].Trim();
        }
        else if (itemArr[0].Trim().ToLower().Equals("port"))
        {
          int port = 0;
          if (int.TryParse(itemArr[1].Trim(), out port))
          {
            Port = port;
          }
          else
          {
            throw new PDBException(PDBErrorCode.PdbE_INVALID_PARAM);
          }
        }
        else if (itemArr[0].Trim().ToLower().Equals("username"))
        {
          UserName = itemArr[1].Trim();
        }
        else if (itemArr[0].Trim().ToLower().Equals("password"))
        {
          Password = itemArr[1].Trim();
        }
      }

      if (string.IsNullOrEmpty(HostName) || Port <= 0 || Port >= 65536
        || string.IsNullOrEmpty(UserName) || string.IsNullOrEmpty(Password))
      {
        throw new PDBException(PDBErrorCode.PdbE_INVALID_PARAM);
      }

      this.ConnStr = connStr;
      sock = null;
    }
    ~PDBConnection()
    {
      this.Close();
    }

    /// <summary>
    /// 打开PinDB数据库连接
    /// </summary>
    public void Open()
    {
      this.Close();
      sock = new TcpClient(HostName, Port);
      sock.SendBufferSize = (256 * 1024);
      sock.ReceiveBufferSize = (64 * 1024);
      sock.NoDelay = true;
      sock.Client.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.KeepAlive, true);

      try
      {
        Login();
      }
      catch (Exception ex)
      {
        this.Close();
        throw ex;
      }

    }

    /// <summary>
    /// 关闭PinDB数据库连接
    /// </summary>
    public void Close()
    {
      if (sock != null)
      {
        sock.Close();
      }
      sock = null;
    }

    internal void Request(byte[] buf)
    {
      if (sock == null)
        throw new PDBException(PDBErrorCode.PdbE_NET_ERROR, "网络错误：未连接");

      sock.GetStream().Write(buf, 0, buf.Length);
    }

    internal byte[] Recv()
    {
      if (sock == null)
        throw new PDBException(PDBErrorCode.PdbE_NET_ERROR, "网络错误：未连接");

      int totalLen = 0;

      byte[] headBuf = new byte[ProtoHeader.kHeadLen];

      while (totalLen < ProtoHeader.kHeadLen)
      {
        int tmpLen = ProtoHeader.kHeadLen - totalLen;
        int recvLen = sock.GetStream().Read(headBuf, totalLen, tmpLen);
        totalLen += recvLen;
      }

      ProtoHeader proHdr = new ProtoHeader(headBuf);
      UInt32 headCrc = proHdr.GetHeadCrc();
      UInt32 tmpHeadCrc = CRC.Crc32(headBuf, 0, 60);
      if (headCrc != tmpHeadCrc)
      {
        throw new PDBException(PDBErrorCode.PdbE_NET_ERROR, "报文错误：回复报文中，回复头校验错误");
      }

      int bodyLen = (int)proHdr.GetBodyLen();

      if (bodyLen == 0)
      {
        return headBuf;
      }

      byte[] packetBuf = new byte[ProtoHeader.kHeadLen + bodyLen];
      Array.Copy(headBuf, 0, packetBuf, 0, ProtoHeader.kHeadLen);
      totalLen = 0;
      while (totalLen < bodyLen)
      {
        int tmpLen = bodyLen - totalLen;
        int recvLen = sock.GetStream().Read(packetBuf, (ProtoHeader.kHeadLen + totalLen), tmpLen);
        totalLen += recvLen;
      }

      UInt32 bodyCrc = proHdr.GetBodyCrc();

      UInt32 tmpBodyCrc = CRC.Crc32(packetBuf, ProtoHeader.kHeadLen, bodyLen);
      if (tmpBodyCrc != bodyCrc)
      {
        throw new PDBException(PDBErrorCode.PdbE_PACKET_ERROR, "报文错误：回复报文中，回复头校验错误");
      }

      return packetBuf;
    }

    /// <summary>
    /// 创建命令对象
    /// </summary>
    /// <returns></returns>
    public PDBCommand CreateCommand()
    {
      return new PDBCommand(this);
    }

    public void Dispose()
    {
      this.Close();
    }

    private void Login()
    {
      byte[] loginPacket = new byte[ProtoHeader.kHeadLen + kUserNameLen + kPwdLen];
      ProtoHeader proHdr = new ProtoHeader(loginPacket);

      Array.Clear(loginPacket, 0, (loginPacket.Length));
      
      byte[] nameArray = System.Text.Encoding.Default.GetBytes(UserName);
      byte[] pwdArray = System.Text.Encoding.Default.GetBytes(Password);

      if (nameArray.Length >= kUserNameLen)
      {
        throw new PDBException(PDBErrorCode.PdbE_INVALID_USER_NAME, "非法的用户名：用户名不合法");
      }
      for (int i = 0; i < nameArray.Length; i++)
      {
        loginPacket[ProtoHeader.kHeadLen + i] = nameArray[i];
      }
      UInt64 pwd64 = CRC.Crc64(pwdArray);
      UInt32 pwd32 = (UInt32)(pwd64 & 0xFFFFFFFF);
      IntTool.PutUint32(loginPacket, (ProtoHeader.kHeadLen + kUserNameLen), pwd32);

      UInt32 bodyCrc = CRC.Crc32(loginPacket, ProtoHeader.kHeadLen, (kUserNameLen + kPwdLen));

      proHdr.SetVersion(ProtoHeader.kProtoVersion);
      proHdr.SetMethod(ProtoHeader.MethodCmdLoginReq);
      proHdr.SetBodyLen((uint)(kUserNameLen + kPwdLen));
      proHdr.SetReturnVal(0);
      proHdr.SetRecordCnt((uint)0);
      proHdr.SetBodyCrc(bodyCrc);
      proHdr.SetHeadCrc();

      Request(loginPacket);

      byte[] resBuf = Recv();
      ProtoHeader resHdr = new ProtoHeader(resBuf);
      if (resHdr.GetMethod() != ProtoHeader.MethodCmdLoginRep)
      {
        throw new PDBException(PDBErrorCode.PdbE_PACKET_ERROR);
      }
      if (resHdr.GetReturnVal() != 0)
      {
        throw new PDBException(PDBErrorCode.PdbE_NET_ERROR, "网络错误：登录失败," + resHdr.GetReturnVal().ToString());
      }

    }

    private string ConnStr { get; set; }
    private string HostName { get; set; }
    private int Port { get; set; }
    private string UserName { get; set; }
    private string Password { get; set; }


  }
}

File: src/pinusdb/connector/PDBJDBC/src/cn/pinusdb/jdbc/PDBConnection.java
package cn.pinusdb.jdbc;

import java.sql.Array;
import java.sql.Blob;
import java.sql.CallableStatement;
import java.sql.Clob;
import java.sql.DatabaseMetaData;
import java.sql.NClob;
import java.sql.PreparedStatement;
import java.sql.SQLClientInfoException;
import java.sql.SQLException;
import java.sql.SQLWarning;
import java.sql.SQLXML;
import java.sql.Savepoint;
import java.sql.Statement;
import java.sql.Struct;
import java.util.Map;
import java.util.Properties;
import java.util.concurrent.Executor;
import java.io.IOException;

import java.net.InetSocketAddress;
import java.net.Socket;

public class PDBConnection implements java.sql.Connection {

	private int serPort_;
	private String serHost_;
	private String userName_;
	private String userPwd_;
	private Socket socket_ = null;
	
	private final int kUserNameLen = 48;
	private final int kPwdLen = 4;
	private final int kDefaultTimeOut = 65000;
	
	PDBConnection(String server, int port, String user, String pwd) 
		throws SQLException
	{
		this.serHost_ = server;
		this.serPort_ = port;
		this.userName_ = user;
		this.userPwd_ = pwd;
		if (serHost_.isEmpty() || userName_.isEmpty() || userPwd_.isEmpty() 
			|| serPort_ <= 0 || serPort_ >= 65536)
		{
			throw new IllegalArgumentException();
		}
		
		open();
	}
	
	public void open() throws SQLException
	{
		socket_ = new Socket();
		try {
			socket_.setReuseAddress(true);
			socket_.setKeepAlive(true);
			socket_.setTcpNoDelay(true);
			socket_.setSoLinger(true, 0);
			socket_.setSoTimeout(kDefaultTimeOut);
			socket_.connect(new InetSocketAddress(serHost_, serPort_));
		} catch (IOException e) {
			throw new SQLException("网络错误", "58005", PDBErrCode.PdbE_NET_ERROR);
		}
		login();
	}
	
    void request(byte[] buf) throws SQLException {
		if (socket_ == null)
			throw new SQLException("网络错误：未连接", "58005", PDBErrCode.PdbE_NET_ERROR);
		
		try {
			socket_.getOutputStream().write(buf, 0, buf.length);
		} catch (IOException e) {
			throw new SQLException(e.getMessage(), "58005", PDBErrCode.PdbE_NET_ERROR);
		}
	}
	
    byte[] recv() throws SQLException {
		if (socket_ == null)
			throw new SQLException("网络错误：未连接", "58005", PDBErrCode.PdbE_NET_ERROR);
		
		int totalLen = 0;
		int headLen = ProtoHeader.getHeadLen();
		byte[] headBuf = new byte[headLen];
		while(totalLen < headLen) {
			int tmpLen = headLen - totalLen;
			int recvLen = 0;
			try {
				recvLen = socket_.getInputStream().read(headBuf, totalLen, tmpLen);
			} catch (IOException e) {
				throw new SQLException(e.getMessage(), "58005", PDBErrCode.PdbE_NET_ERROR);
			}
			if (recvLen < 0) {
				throw new SQLException(PDBErrCode.errMsg(PDBErrCode.PdbE_NET_ERROR), 
					"58005", PDBErrCode.PdbE_NET_ERROR);
			}
			totalLen += recvLen;
		}
		
		ProtoHeader proHdr = new ProtoHeader(headBuf);
		int headCrc = proHdr.getHeadCrc();
		int tmpHeadCrc = CRC.crc32(headBuf, 0, 60);
		if (headCrc != tmpHeadCrc) {
			throw new SQLException("报文错误,回复报文中,报文头校验错误", "58005", PDBErrCode.PdbE_PACKET_ERROR);
		}
		
		int bodyLen = proHdr.getBodyLen();
		if (bodyLen == 0) {
			return headBuf;
		}
		
		byte[] packetBuf = new byte[ProtoHeader.getHeadLen() + bodyLen];
		System.arraycopy(headBuf, 0, packetBuf, 0, headBuf.length);
		totalLen = 0;
		while (totalLen < bodyLen) {
			int tmpLen = bodyLen - totalLen;
			int recvLen = 0;
			try {
				recvLen = socket_.getInputStream().read(packetBuf, 
					(ProtoHeader.getHeadLen() + totalLen), tmpLen);
			} catch (IOException e) {
				throw new SQLException("网络错误,读取报文失败", "58005", PDBErrCode.PdbE_NET_ERROR);
			}
			totalLen += recvLen;
		}
		
		int bodyCrc = proHdr.getBodyCrc();
		int tmpBodyCrc = CRC.crc32(packetBuf, ProtoHeader.getHeadLen(), bodyLen);
		if (tmpBodyCrc != bodyCrc) {
			throw new SQLException("报文错误,回复报文中,报文体校验失败", "58005", PDBErrCode.PdbE_PACKET_ERROR);
		}
		
		return packetBuf;
	}
	
	public void dispose() throws SQLException {
		this.close();
	}
	
	private void login() throws SQLException {
		byte[] loginPacket = new byte[ProtoHeader.getHeadLen() + kUserNameLen + kPwdLen];
		ProtoHeader proHdr = new ProtoHeader(loginPacket);
		
		for (int i = 0; i < loginPacket.length; i++) {
			loginPacket[i] = 0;
		}
		
		byte[] nameArr = this.userName_.getBytes();
		byte[] pwdArr = this.userPwd_.getBytes();
		if (nameArr.length >= kUserNameLen) {
			throw new SQLException("错误的参数: 用户名不合法", "58005", PDBErrCode.PdbE_INVALID_USER_NAME);
		}
		
		System.arraycopy(nameArr, 0, loginPacket, ProtoHeader.getHeadLen(), nameArr.length);
		int pwd32 = CRC.crc32(pwdArr, 0, pwdArr.length);
		IntTool.setInt32(loginPacket, (ProtoHeader.getHeadLen() + kUserNameLen), pwd32);
		int bodyCrc = CRC.crc32(loginPacket, ProtoHeader.getHeadLen(), (kUserNameLen + kPwdLen));
		
		proHdr.setVersion(ProtoHeader.getProtoVersion());
		proHdr.setMethod(ProtoHeader.getMethodCmdLoginReq());
		proHdr.setBodyLen((kUserNameLen + kPwdLen));
		proHdr.setBodyCrc(bodyCrc);
		proHdr.updateHeadCrc();
		
		request(loginPacket);
		
		byte[] recvBuf = recv();
		ProtoHeader recvHdr = new ProtoHeader(recvBuf);
		if (recvHdr.getMethod() != ProtoHeader.getMethodCmdLoginRep()) {
			throw new SQLException("报文错误,回复方法错误", "58005", PDBErrCode.PdbE_PACKET_ERROR);
		}
		
		int loginRet = recvHdr.getReturnVal();
		if (loginRet != PDBErrCode.PdbE_OK) {
			throw new SQLException(PDBErrCode.errMsg(loginRet), "58005", loginRet);
		}
	}
	
	public PDBCommand createCommand() {
		return new PDBCommand(this);
	}
	
	@Override
	public <T> T unwrap(Class<T> iface) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public boolean isWrapperFor(Class<?> iface) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public Statement createStatement() throws SQLException {
		return new PDBStatement(this);
	}

	@Override
	public PreparedStatement prepareStatement(String sql) throws SQLException {
		return new PDBPreStatement(this, sql);
	}

	@Override
	public CallableStatement prepareCall(String sql) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public String nativeSQL(String sql) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void setAutoCommit(boolean autoCommit) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public boolean getAutoCommit() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void commit() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void rollback() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void close() throws SQLException {
		try {
			socket_.close();
		} catch (IOException e) {
			throw new SQLException("网络错误", "58005", PDBErrCode.PdbE_NET_ERROR);
		}
	}

	@Override
	public boolean isClosed() throws SQLException {
		return socket_.isClosed();
	}

	@Override
	public DatabaseMetaData getMetaData() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void setReadOnly(boolean readOnly) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public boolean isReadOnly() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void setCatalog(String catalog) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public String getCatalog() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void setTransactionIsolation(int level) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public int getTransactionIsolation() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public SQLWarning getWarnings() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void clearWarnings() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public Statement createStatement(int resultSetType, int resultSetConcurrency) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public PreparedStatement prepareStatement(String sql, int resultSetType, int resultSetConcurrency)
			throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public CallableStatement prepareCall(String sql, int resultSetType, int resultSetConcurrency) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public Map<String, Class<?>> getTypeMap() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void setTypeMap(Map<String, Class<?>> map) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void setHoldability(int holdability) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public int getHoldability() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public Savepoint setSavepoint() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public Savepoint setSavepoint(String name) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void rollback(Savepoint savepoint) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void releaseSavepoint(Savepoint savepoint) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public Statement createStatement(int resultSetType, int resultSetConcurrency, int resultSetHoldability)
			throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public PreparedStatement prepareStatement(String sql, int resultSetType, int resultSetConcurrency,
			int resultSetHoldability) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public CallableStatement prepareCall(String sql, int resultSetType, int resultSetConcurrency,
			int resultSetHoldability) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public PreparedStatement prepareStatement(String sql, int autoGeneratedKeys) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public PreparedStatement prepareStatement(String sql, int[] columnIndexes) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public PreparedStatement prepareStatement(String sql, String[] columnNames) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public Clob createClob() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public Blob createBlob() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public NClob createNClob() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public SQLXML createSQLXML() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public boolean isValid(int timeout) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void setClientInfo(String name, String value) throws SQLClientInfoException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void setClientInfo(Properties properties) throws SQLClientInfoException {
		throw new UnsupportedOperationException("不支持的方法"); 
		
	}

	@Override
	public String getClientInfo(String name) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public Properties getClientInfo() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public Array createArrayOf(String typeName, Object[] elements) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public Struct createStruct(String typeName, Object[] attributes) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void setSchema(String schema) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public String getSchema() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void abort(Executor executor) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public void setNetworkTimeout(Executor executor, int milliseconds) throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}

	@Override
	public int getNetworkTimeout() throws SQLException {
		throw new UnsupportedOperationException("不支持的方法"); 
	}
}
